src/Kestrel.Core/Internal/Http/HttpProtocol.cs

```diff
@@ -533,7 +533,7 @@ public void OnHeader(Span<byte> name, Span<byte> value)
                             await ProduceEnd();
 
                             // ForZeroContentLength does not complete the reader nor the writer
-                            if (!messageBody.IsEmpty && _keepAlive)
```

Comment:
I'm not sure I like setting the IsEmpty property true for upgraded connections since it feels a middle misleading. Maybe we can leave the ForUpgrade class as is and change this condition to `if (!messageBody.IsEmpty && !messageBody.RequestUpgrade)` to make things more explicit.

Reply:
The reason why there is no need to consume the body is only because there is no request body.
The upgrade request has always no body as far as I understand, in fact there is a check for the `Content-Length` to be zero [here](https://github.com/alefranz/KestrelHttpServer/blob/de2ecaa53051e669fc4ba9309b9646c30320b826/src/Kestrel.Core/Internal/Http/Http1MessageBody.cs#L233-L238),
so to me it makes sense to set the body as empty. What do you think?

However in the [RFC](https://tools.ietf.org/html/rfc7230#section-6.7) there is no mention that an upgrade request must have no payload (but I'm no expert on this), so I guess this behaviour can potentially change in the feature, but if then an upgrade request can have a body, it will makes sense to always ensure the body is consumed.

Reply:
@alefranz You've convinced me. Technically we're not dealing with a message body for upgraded requests.